flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -434,6 +434,13 @@ protected void beforeInvoke() throws Exception {
 			// so that we avoid race conditions in the case that initializeState()
 			// registers a timer, that fires before the open() is called.
 			operatorChain.initializeStateAndOpenOperators(createStreamTaskStateInitializer());
+
+			ResultPartitionWriter[] writers = getEnvironment().getAllWriters();
+			if (writers != null) {
+				for (ResultPartitionWriter writer : writers) {
+					writer.initializeState();
+				}
+			}
```

Comment:
Should we close the reader here? Or it will be used for InputChannels later?

Reply:
The reader would also be used for input side for recovery in another PR. Have you considered when to close it in your PR? The `ChannelStateReader` is a task-level component and passed into `Environment` to be used anywhere in theory. Input and output sides are the users for reader ATM, but they should not be aware whether there are other users. So they should not close it after using . The proper way might consider its lifecycle together with task's lifecycle. WDYT?

Reply:
Yes, I fully agree with what you point. In my PR, I'm closing it in `Task.releaseResources`. Just wanted to make sure that we can't close it sooner.

In `ChannelStreamReaderImpl`, the underlying streams are closed as soon as all the data is read from them, so it shouldn't be an issue.